src/hooks/chart.ts

```diff
@@ -817,6 +874,144 @@ const parseBarResponse = (trace: Trace, plot: Plot, responseData: ResponseData)
   return result;
 };
 
+
+/**
+ * Parses response data obtained for every trace within the plot
+ *
+ * @param trace current trace the plot config
+ * @param plot plot config
+ * @param responseData data received from request to be parsed
+ * @returns
+ */
+const parseHeatmapResponse = (trace: Trace, plot: Plot, responseData: ResponseData) => {
+  const result: Partial<TraceConfig> & Partial<PlotlyPlotData> & PlotResultData & ClickableLinks & HeatmapZData = {
+    ...trace,
+    type: plot.plot_type,
+    x: [], // x data
+    y: [], // y data
+    z: [], // z data
+    text: [], // text data
+    legend_clickable_links: [], // array of links for when clicking legend
+    graphic_clickable_links: [], // array of links for when clicking graph
+  };
+  let yIndex = 0;
+  let layoutParams={};
+  let longestXTick='';
+  let longestYTick='';
+  const { config, plotly } = plot;
+  const { xaxis, yaxis, format_data_x = false, format_data_y = false } = config;
+  responseData.forEach((item: any, i:number) => {
+    updateWithTraceColData(result, trace, item, i);
+    // Add the y values for the heatmap plot
+    trace?.y_col?.forEach((colName: string, i: number) => {
+      const value = getValue(item, colName, yaxis, format_data_y, plot);
+      // Adds the y value for the heatmap plot if it is not added yet in y array
+      if (result.y.indexOf(value.toString()) < 0) {
+        result.y.push(value.toString());
+        if(item[colName].toString().length > longestYTick?.length){
+          longestYTick=item[colName].toString();
+        }
+        result.z.push([]);
+      }
+      // Fetching the index of added y value to add corresponding z value
+      yIndex = result.y.indexOf(value.toString())
+    });
+    // Add the z values for the heatmap plot based on y val's index
+    trace?.z_col?.forEach((colName: string) => {
+      const zValue = item[colName];
+      result?.z[yIndex]?.push(zValue);
+
+    });
+    // Add the x values for the heatmap plot if its not added yet
+    trace?.x_col?.forEach((colName: string) => {
+      const value = getValue(item, colName, xaxis, format_data_x, plot);
+      if (result.x.indexOf(value.toString()) < 0) {
+        if(item[colName].toString().length > longestXTick?.length){
+          longestXTick=item[colName].toString();
+        }
+        result.x.push(value.toString());
+      }
+    });
+
+    // Getting the longest x tick in the given data to determine margin and height values in getHeatmapLayoutParams function
+    const inputParams = {
+      width: typeof plotly?.layout.width !== 'undefined' ? plotly?.layout.width : 1200,
+      xTickAngle: typeof plotly?.layout.xaxis?.tickangle !== 'undefined' ? plotly?.layout.xaxis?.tickangle : 50,
+    }
+    layoutParams = getHeatmapLayoutParams(inputParams, longestXTick?.length, longestYTick?.length, result.y?.length);
```

Comment:
`layoutParams` is being calculated for each `row` in `responseData`. Let's move this outside of `responseData.forEach(...)` so it is only done once after we create the `result` and calculate the longest x and y ticks.